### PR TITLE
Add Caching for most file system interactions

### PR DESCRIPTION
**Pull request by JMPJNS**
Cache Everything that is not a binary file in a keyv store

### CODE REVIEW

File: src/modules/filesystem.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import { dataDirectory, imagesDirectory } from '../config';
+import keyv from 'keyv';
+import { promises as fs, existsSync } from 'fs';
+import mimeTypes from 'mime-types';
+import path from 'path';
+import sharp from 'sharp';
+
+const cache = new keyv();
+
+export async function getTypes(): Promise<string[]> {
```
Comment (RunAge) [inline src/modules/filesystem.ts:10]: This should be an enum.

File: src/modules/filesystem.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    const englishExists = existsSync(englishPath);
+    if (englishExists) errorMessage += `, language en would exist`;
+
+    throw new Error(errorMessage);
+  }
+
+  const file = await fs.readFile(filePath);
+  try {
+    const entity = JSON.parse(file.toString('utf-8'));
+    await await cache.set(`data-${type}-${id}-${lang}`, entity);
```
Comment (RunAge) [inline src/modules/filesystem.ts:65]: Is it possible to cache it for contain amount of time? I think its not needed to cache it indefinitely but store only frequently used data to reduce ram usage.

File: src/modules/filesystem.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  type: string,
+): Promise<string[] | null> {
+  const found = await cache.get(`data-${type}`);
+  if (found) return found;
+
+  const exists = existsSync(dataDirectory(type));
+  if (!exists) throw new Error(`Type ${type} not found`);
+
+  const entities = await fs.readdir(dataDirectory(type));
+  await cache.set(`data-${type}`, entities);
```
Comment (RunAge) [inline src/modules/filesystem.ts:32]: Is it possible to cache it for contain amount of time? I think its not needed to cache it indefinitely but store only frequently used data to reduce ram usage.

File: src/modules/filesystem.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  const found = await cache.get(`image-${type}-${id}`);
+  if (found) return found;
+
+  const filePath = path.join(imagesDirectory(type), id).normalize();
+  if (!existsSync(filePath)) {
+    throw new Error(`Entity ${type}/${id} doesn't exist`);
+  }
+
+  const images = await fs.readdir(filePath);
+  await cache.set(`image-${type}-${id}`, images);
```
Comment (RunAge) [inline src/modules/filesystem.ts:88]: Is it possible to cache it for contain amount of time? I think its not needed to cache it indefinitely but store only frequently used data to reduce ram usage.

File: src/modules/filesystem.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  console.log(`added image-${type}-${id} to cache`);
+  return images;
+}
+
+export async function getImage(type: string, id: string, image: string) {
```
Comment (RunAge) [inline src/modules/filesystem.ts:93]: I think cache also should be applied here.

**Review by RunAge — CHANGES_REQUESTED** [review]

File: src/modules/filesystem.ts (hunk 2 above)
Comment (JMPJNS) [reply src/modules/filesystem.ts:65]: yes it is, keyv supports a third parameter, which is the time you want to store it

File: src/modules/filesystem.ts (hunk 5 above)
Comment (JMPJNS) [reply src/modules/filesystem.ts:93]: i don't think that we should cache binary data

File: src/modules/filesystem.ts (hunk 1 above)
Comment (JMPJNS) [reply src/modules/filesystem.ts:10]: the types get evaluated dynamically by the contents of the /assets folder, not sure how we would solve that with an enum without having to change the source code to add new types

File: src/modules/filesystem.ts (hunk 5 above)
Comment (RunAge) [reply src/modules/filesystem.ts:93]: cache should be used to reduce reading from drive frequently used data, because it is unnecessary work and slow

**Review by NurMarvin — APPROVED** [review]

**Pull request merged**